crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4082,27 +4142,35 @@ impl<'db> TypeInferenceBuilder<'db> {
     ) -> Type<'db>
     where
         Iterator: IntoIterator<Item = Item>,
-        F: Fn(&mut Self, Item) -> Type<'db>,
+        F: Fn(&mut Self, Item) -> (Type<'db>, TextRange),
     {
         let mut done = false;
         let db = self.db();
 
         let elements = operations
             .into_iter()
             .with_position()
-            .map(|(position, ty)| {
-                let ty = infer_ty(self, ty);
-
-                if done {
-                    return Type::Never;
-                }
+            .map(|(position, item)| {
+                let (ty, range) = infer_ty(self, item);
 
                 let is_last = matches!(
                     position,
                     itertools::Position::Last | itertools::Position::Only
                 );
 
-                match (ty.bool(db), is_last, op) {
+                let truthiness = ty.try_bool(self.db()).unwrap_or_else(|err| {
+                    if !is_last {
+                        err.report_diagnostic(&self.context, range);
+                    }
+
+                    err.fallback_truthiness()
+                });
+
+                if done {
+                    return Type::Never;
+                }
+
+                match (truthiness, is_last, op) {
```

Comment:
I think this logic and the match below could be adjusted so we can avoid the unnecessary `try_bool` call on the last comparison entirely, instead of just skipping the diagnostic from it. Skipping the diagnostic is a tell that we are calling `try_bool` when we shouldn't need to because runtime doesn't.

If `is_last`, we always return `ty` and we don't actually care about `truthiness`.

I haven't checked the surrounding logic to see whether we enter this path even on single (non chained) comparisons, but if we do, fixing this could result in significantly fewer useless `try_bool` calls in real codebases.